File: agriculture_scraping/scripts/scraping/meteo_scraper.py
# -*- coding: utf-8 -*-
"""
INM (Institut National de la Météorologie) - meteo.tn
Tentative de récupération des données climatiques (températures, pluies) pour la Tunisie.
Le site meteo.tn est principalement informatif; les données historiques peuvent
être limitées. On complète avec Open-Meteo pour les séries temporelles.
"""

from pathlib import Path

import pandas as pd
import requests
from bs4 import BeautifulSoup

INM_BASE = "https://www.meteo.tn"
# Page des données / climat (à adapter si le site propose des exports)
INM_FR = "https://www.meteo.tn/fr"


def scrape_inm_links() -> list[dict]:
    """
    Parcourt le site INM pour trouver des liens vers des données
    (fichiers CSV, Excel, ou pages avec tableaux).
    """
    found = []
    try:
        r = requests.get(INM_FR, timeout=15, headers={"User-Agent": "Mozilla/5.0 (compatible; AgriBot/1.0)"})
        r.raise_for_status()
        soup = BeautifulSoup(r.text, "lxml")
        for a in soup.find_all("a", href=True):
            href = a.get("href", "")
            text = (a.get_text() or "").strip().lower()
            if any(k in href.lower() or k in text for k in ["donnee", "data", "climat", "historique", "csv", "xls", "export"]):
                if not href.startswith("http"):
                    href = f"{INM_BASE.rstrip('/')}/{href.lstrip('/')}"
                found.append({"url": href, "text": text[:100]})
    except Exception as e:
        print(f"INM scrape: {e}")
    return found


def scrape_inm_tables(output_dir: str | Path) -> pd.DataFrame:
    """
    Tente d'extraire des tableaux HTML depuis les pages INM (températures, pluies).
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    all_tables = []
    try:
        r = requests.get(INM_FR, timeout=15, headers={"User-Agent": "Mozilla/5.0 (compatible; AgriBot/1.0)"})
        r.raise_for_status()
        soup = BeautifulSoup(r.text, "lxml")
        for i, table in enumerate(soup.find_all("table")):
            try:
                df = pd.read_html(str(table), encoding="utf-8")[0]
                if df.shape[0] > 1 and df.shape[1] > 1:
                    df["_source_page"] = "meteo_tn_fr"
                    df["_table_index"] = i
                    all_tables.append(df)
            except Exception:
                continue
    except Exception as e:
        print(f"INM tables: {e}")

    if all_tables:
        combined = pd.concat(all_tables, ignore_index=True)
        out_path = output_dir / "meteo_tn_tables.csv"
        combined.to_csv(out_path, index=False, encoding="utf-8-sig")
        return combined
    return pd.DataFrame()


def collect_inm_data(output_dir: str | Path) -> pd.DataFrame:
    """
    Point d'entrée: récupère les données disponibles sur meteo.tn.
    Si peu de données sont trouvées, le pipeline utilisera Open-Meteo pour le climat.
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    links = scrape_inm_links()
    df_tables = scrape_inm_tables(output_dir)

    if links:
        (output_dir / "meteo_tn_links.txt").write_text(
            "\n".join(f"{x['url']}\t{x['text']}" for x in links[:50]),
            encoding="utf-8",
        )
    return df_tables

File: R/scraping/meteo_scraper.R
# INM (Institut National de la Météorologie) - meteo.tn
# Récupération des données climatiques (températures, pluies) pour la Tunisie.

library(httr)
library(rvest)
library(dplyr)
library(readr)

INM_BASE <- "https://www.meteo.tn"
INM_FR   <- "https://www.meteo.tn/fr"

.USER_AGENT <- "Mozilla/5.0 (compatible; AgriBot/1.0)"


scrape_inm_links <- function() {
  found <- list()
  tryCatch({
    resp <- httr::GET(INM_FR, httr::timeout(15), httr::user_agent(.USER_AGENT))
    if (httr::status_code(resp) != 200) return(found)

    page  <- rvest::read_html(httr::content(resp, as = "text", encoding = "UTF-8"))
    nodes <- rvest::html_nodes(page, "a[href]")
    hrefs <- rvest::html_attr(nodes, "href")
    texts <- trimws(rvest::html_text(nodes))

    keywords <- c("donnee", "data", "climat", "historique", "csv", "xls", "export")
    for (i in seq_along(hrefs)) {
      href <- hrefs[i]
      text <- tolower(texts[i])
      href_lower <- tolower(href %||% "")
      if (any(sapply(keywords, function(k) grepl(k, href_lower) || grepl(k, text)))) {
        if (!startsWith(href, "http")) {
          href <- paste0(sub("/$", "", INM_BASE), "/", sub("^/", "", href))
        }
        found[[length(found) + 1]] <- list(url = href, text = substr(text, 1, 100))
      }
    }
  }, error = function(e) cat(sprintf("INM scrape: %s\n", conditionMessage(e))))
  found
}

`%||%` <- function(x, y) if (is.null(x) || length(x) == 0) y else x


scrape_inm_tables <- function(output_dir) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  all_tables <- list()
  tryCatch({
    resp <- httr::GET(INM_FR, httr::timeout(15), httr::user_agent(.USER_AGENT))
    if (httr::status_code(resp) != 200) return(data.frame())

    page   <- rvest::read_html(httr::content(resp, as = "text", encoding = "UTF-8"))
    tables <- rvest::html_table(page, fill = TRUE)

    for (i in seq_along(tables)) {
      tbl <- tables[[i]]
      if (nrow(tbl) > 1 && ncol(tbl) > 1) {
        tbl$`_source_page`  <- "meteo_tn_fr"
        tbl$`_table_index`  <- i
        all_tables[[length(all_tables) + 1]] <- tbl
      }
    }
  }, error = function(e) cat(sprintf("INM tables: %s\n", conditionMessage(e))))

  if (length(all_tables) > 0) {
    combined <- dplyr::bind_rows(all_tables)
    readr::write_csv(combined, file.path(output_dir, "meteo_tn_tables.csv"))
    return(invisible(combined))
  }
  data.frame()
}


collect_inm_data <- function(output_dir) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  links    <- scrape_inm_links()
  df_tables <- scrape_inm_tables(output_dir)

  if (length(links) > 0) {
    link_lines <- sapply(links[seq_len(min(50, length(links)))], function(x) {
      paste0(x$url, "\t", x$text)
    })
    writeLines(link_lines, file.path(output_dir, "meteo_tn_links.txt"))
  }

  invisible(df_tables)
}


if (!interactive()) {
  args <- commandArgs(trailingOnly = TRUE)
  out  <- if (length(args) >= 1) args[1] else "data"
  df   <- collect_inm_data(out)
  cat(sprintf("Lignes INM: %d\n", nrow(df)))
}
